cpp/daal/src/algorithms/k_nearest_neighbors/kdtree_knn_classification_train_dense_default_impl.i

```diff
@@ -61,39 +61,63 @@ using namespace kdtree_knn_classification::internal;
 template <typename T, CpuType cpu>
 class Queue
 {
+    static const size_t defaultSize = 4;
+
 public:
-    Queue() : _data(nullptr) {}
+    Queue() : _data(nullptr), _first(0), _last(0), _count(0), _capacity(0) {}
 
-    ~Queue()
-    {
-        services::daal_free(_data);
-        _data = nullptr;
-    }
+    ~Queue() { clear(); }
+
+    Queue(const Queue &)             = delete;
+    Queue & operator=(const Queue &) = delete;
 
     bool init(size_t size)
     {
         clear();
-        _first = _count = 0;
-        _last = _sizeMinus1 = (_size = size) - 1;
-        return ((_data = static_cast<T *>(service_malloc<T, cpu>(size * sizeof(T)))) != nullptr);
+        if (size == 0)
+        {
+            return false;
+        }
+
+        _data = static_cast<T *>(service_malloc<T, cpu>(size));
+        if (!_data)
+        {
+            return false;
+        }
+
+        _capacity = size;
+        _first = _last = _count = 0;
+        return true;
     }
 
     void clear()
     {
-        daal_free(_data);
-        _data = nullptr;
+        if (_data)
+        {
+            daal::services::internal::service_free<T, cpu>(_data);
+            _data = nullptr;
+        }
+        _first = _last = _count = _capacity = 0;
     }
 
-    DAAL_FORCEINLINE void push(const T & value)
+    void reset() { _first = _last = _count = 0; }
+
+    void push(const T & value)
     {
-        _data[_last = (_last + 1) & _sizeMinus1] = value;
+        if (_count >= _capacity)
+        {
+            grow();
+        }
+
+        _data[_last] = value;
+        ++_last;
         ++_count;
     }
 
-    DAAL_FORCEINLINE T pop()
+    T pop()
```

Comment:
Looks like `--_last;` is missing.
Maybe that's the reason why `grow()` started to be required?

Reply:
Is it necessary to call `--last`. I move the index of the first element, but I shouldnt touch the last one. _last indicates the next available position for insertion, not the index of the last element